velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -1042,6 +1044,123 @@ TpchPlan TpchQueryBuilder::getQ19Plan() const {
+  std::vector<std::string> supplierColumns = {
+      "s_nationkey", "s_suppkey", "s_name", "s_address"};
+  std::vector<std::string> partsuppColumns = {
+      "ps_partkey", "ps_suppkey", "ps_availqty"};
+  std::vector<std::string> nationColumns = {"n_nationkey", "n_name"};
+
+  auto lineitemSelectedRowType = getRowType(kLineitem, lineitemColumns);
+  const auto& lineitemFileColumns = getFileColumnNames(kLineitem);
+  auto partSelectedRowType = getRowType(kPart, partColumns);
+  const auto& partFileColumns = getFileColumnNames(kPart);
+  auto supplierSelectedRowType = getRowType(kSupplier, supplierColumns);
+  const auto& supplierFileColumns = getFileColumnNames(kSupplier);
+  auto partsuppSelectedRowType = getRowType(kPartsupp, partsuppColumns);
+  const auto& partsuppFileColumns = getFileColumnNames(kPartsupp);
+  auto nationSelectedRowType = getRowType(kNation, nationColumns);
+  const auto& nationFileColumns = getFileColumnNames(kNation);
+
+  const std::string shipDateFilter = formatDateFilter(
+      "l_shipdate", lineitemSelectedRowType, "'1994-01-01'", "'1994-12-31'");
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId lineitemScanNodeId;
+  core::PlanNodeId partScanNodeId;
+  core::PlanNodeId supplierScanNodeId;
+  core::PlanNodeId partsuppScanNodeId;
+  core::PlanNodeId nationScanNodeId;
+
+  auto part = PlanBuilder(planNodeIdGenerator)
+                  .tableScan(
+                      kPart,
+                      partSelectedRowType,
+                      partFileColumns,
+                      {},
+                      "p_name like 'forest%'")
+                  .capturePlanNodeId(partScanNodeId)
+                  .planNode();
+
+  auto nation = PlanBuilder(planNodeIdGenerator)
+                    .tableScan(
+                        kNation,
+                        nationSelectedRowType,
+                        nationFileColumns,
+                        {"n_name = 'CANADA'"})
+                    .capturePlanNodeId(nationScanNodeId)
+                    .planNode();
+
+  auto partsuppJoinPart =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kPartsupp, partsuppSelectedRowType, partsuppFileColumns)
+          .capturePlanNodeId(partsuppScanNodeId)
+          .hashJoin(
+              {"ps_partkey"},
+              {"p_partkey"},
+              part,
+              "",
+              {"ps_partkey", "ps_suppkey", "ps_availqty"},
+              core::JoinType::kLeftSemi)
+          .planNode();
+
+  auto lineitemJoinPartsupp =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(
+              kLineitem,
+              lineitemSelectedRowType,
+              lineitemFileColumns,
+              {shipDateFilter})
+          .capturePlanNodeId(lineitemScanNodeId)
+          .partialAggregation(
+              {"l_partkey", "l_suppkey"}, {"sum(l_quantity) AS sum_qty"})
+          .localPartition({})
```

Comment:
Is it necessary to gather all data into a single thread? Can we repartition the data using l_partkey and l_suppkey and continue processing over multiple threads?

Reply:
Thanks for the suggestion, modified accordingly